src/libraries/System.Text.Json/Common/JsonSeparatorNamingPolicy.cs

```diff
@@ -21,149 +28,148 @@ public sealed override string ConvertName(string name)
                 ThrowHelper.ThrowArgumentNullException(nameof(name));
             }
 
-            // Rented buffer 20% longer that the input.
-            int rentedBufferLength = (12 * name.Length) / 10;
-            char[]? rentedBuffer = rentedBufferLength > JsonConstants.StackallocCharThreshold
-                ? ArrayPool<char>.Shared.Rent(rentedBufferLength)
-                : null;
-
-            int resultUsedLength = 0;
-            Span<char> result = rentedBuffer is null
-                ? stackalloc char[JsonConstants.StackallocCharThreshold]
-                : rentedBuffer;
+            return ConvertNameCore(_separator, _lowercase, name);
+        }
 
-            void ExpandBuffer(ref Span<char> result)
-            {
-                char[] newBuffer = ArrayPool<char>.Shared.Rent(result.Length * 2);
+        private static string ConvertNameCore(char separator, bool lowercase, string name)
+        {
+            Debug.Assert(name != null);
 
-                result.CopyTo(newBuffer);
+            char[]? rentedBuffer = null;
 
-                if (rentedBuffer is not null)
-                {
-                    result.Slice(0, resultUsedLength).Clear();
-                    ArrayPool<char>.Shared.Return(rentedBuffer);
-                }
+            // Rented buffer 20% longer that the input.
+            int initialBufferLength = (12 * name.Length) / 10;
+            Span<char> destination = initialBufferLength <= JsonConstants.StackallocCharThreshold
+                ? stackalloc char[JsonConstants.StackallocCharThreshold]
+                : (rentedBuffer = ArrayPool<char>.Shared.Rent(initialBufferLength));
 
-                rentedBuffer = newBuffer;
-                result = rentedBuffer;
-            }
+            ReadOnlySpan<char> chars = name.AsSpan();
```

Comment:
Why is this local needed rather than all uses of `chars` below just using `name`?

Reply:
Looks like a leftover here since the proposed implementation operates on separate characters rather than on words as the original does which brings an issue because characters have one-to-one mapping, but that's false for strings.

Reply:
Just as a reference for all of those who are reviewing the pull request, here are two implementations I made before migrating policies to .NET:

* Without true word segmentation according to the Unicode rules (the current mine and what you have in `main`): https://github.com/YohDeadfall/Yoh.Text.Json.NamingPolicies/blob/6ca9e48cf9bf1dc0d1b15ac1df46bfc02db10557/src/JsonSeparatorNamingPolicy.cs

* With true word segmentation according to the Unicode rules: https://github.com/YohDeadfall/Yoh.Text.Json.NamingPolicies/blob/cfbf445538d4ff8410b8584ab75009d4368777c9/src/JsonSimpleNamingPolicy.cs

The last one was the most correct implementation, but I had to remove Unicode text segmentation due to my plans on merging that work into .NET. It required a bit of side code because there are no text segmentation support (discussed with Levi Broderick on Twitter, he thought that won't find use case) and due to having worse performance than without it while producing same outputs for most important cases.